### DISCUSSION TITLE
Programmatic Access to Server Mod Lists for Streamlining Administration and Marketing at Scale

### DISCUSSION THREAD

**Discussion by TDodgeCo** [gameserverapp/Platform#947] [Suggestions]
## Feature Request Overview
This proposal suggests the implementation of programmatic access to the list of mods installed on servers through the community API. The feature aims to streamline mod management for large-scale server deployments by enabling dynamic retrieval of installed mod lists, improving automation and reducing manual intervention for community administrators.

### Problem Statement
Currently, users have no direct way to access mod lists programmatically, which creates inefficiencies for administrators managing multiple servers with frequent mod updates. The manual process of retrieving mod lists, processing them, and uploading the data to external platforms is time-consuming and prone to errors—especially for environments with a high number of servers, such as clusters running ~40 mods per server. As server fleets grow, this challenge will scale alongside them, making it crucial to introduce a solution that reduces administrative overhead.

#### Related Suggestions 
#718 #658 #782 #342
A few of those go beyond the scope of a a simple `get` request, but align with the general intention of providing users the tools to automate the administration of server mods.

### Proposed Solution
Max's proposed solution is to extend the existing community API to include mod lists per server. This would allow administrators to:
 - Dynamically fetch the mod list for each server via the API.
 - Automate the process of updating mod lists on external platforms, reducing manual labor and error risk.
 - Integrate mod data with other administrative tools for better mod management across clusters.
 - Max suggested that the most logical place to implement this feature would be in the server object returned by the API. Including the mod list within this structure would centralize key server data in a single request, streamlining its integration into existing workflows.

### Alternative Ideas
While adding mod lists to the server object is the proposed solution and I generally agree it would be a fine solution, a few alternative approaches could also be considered to inspire further discussion:

- The option to subscribe to webhook notifications for mod changes, pushing updates when mods are added or removed.
- Add a command to the GSA Discord bot to retrieve and share Mod data. This could then be scraped by our own bot.
- A mod export tool in the dashboard UI, offering an easy, one-click data export for administrators who prefer manual workflows but want a faster solution.

```
[ServerSettings]
ActiveMods=%%MODIDS%%
```
- is currently returned in the `GameUserSettings.ini` response from the config template API, express and bundle the value of that variable with the `GUS` response. This would eliminate the need for manual parsing or post-processing of the returned config data. This isn't the ideal long term solution for my use case, but it would serve as an excellent band aid for the time being.

### Key Benefits
 - Efficiency: Automates mod list retrieval, saving administrators significant time, especially when managing large collections of servers.
 - Scalability: Simplifies the process of maintaining up-to-date mod lists as server fleets expand and configurations change.
 - Flexibility: Provides a programmatic method for integrating mod lists into custom community tools, improving platform adaptability for power users.
 
### My Current Workaround and Planned Stop-Gap Solution
I currently have a very basic PowerShell script that I run manually on my dedicated machines to gather a rough list of mod IDs for my Ark Ascended servers. Moving forward, I'm planning to expand this into a Go or Node.js filesystem server that monitors for changes and sends differential data to my web server for processing.

That said, I'm fairly weak with Docker, Windows Server, and networking. I’m concerned about the potential for conflicts with your Docker setup, particularly if I introduce additional network traffic on port 443 and continually poll the mod directories of my server containers.

Any guidance on whether this might create issues would be greatly appreciated.

### Conclusion
Introducing programmatic access to mod lists through the community API would significantly enhance the administrative experience for users managing multiple servers. This feature aligns with the goals of improving efficiency and scalability, making it an invaluable addition to the platform for those running complex server configurations.

**Comment by babsgosgens** [comment]
Hi,

Re.: Integrate mod data with other administrative tools for better mod management across clusters.

Can you explain a bit what you mean by this?

**Reply by TDodgeCo** [reply]
Yeah, sure thing!

I started building myself a custom community site a few months back. That turned into a custom CMS to build the website with... and by this point has spiraled into a full blown community/marketing/administration/content management platform.  Basically covering any issues I run into that aren't currently handled by Max and GSA. 

Aside from the awesome marketing potential of being able to render dynamic lists of installed mods and their related configs, I need better visibility on on the state of my server's mod lists. It's super time consuming and extremely error prone trying to compare lists by navigating back and forth between the mod tab on the server management page. 

So I've been building it. Should be finished with that piece hopefully by the end of the weekend.

**Comment by mvaessen** [comment]
![Screenshot 2024-10-18 at 17 11 00](https://github.com/user-attachments/assets/23392c6a-5506-43b6-ad61-1b6803b5029f)
![Screenshot 2024-10-18 at 17 22 10](https://github.com/user-attachments/assets/c0932306-3a06-45f7-8fc8-d93f4a64522c)

Planning to push this next week. The mod info is provided for the `/api/v1/server/{id}` endpoint only (community api)

The mod / framework data will be included with the response when adding `mods=1` to the query string: `/api/v1/server/127?mods=1`
The data is cached for 2 minutes, at the endpoint

**Reply by TDodgeCo** [reply]
What a stud! Thank you!

**Reply by mvaessen** [reply]
planning to roll this out later today. See the updated info above for details how to grab this info

**Reply by mvaessen** [reply]
This is now available
